niworkflows/utils/spaces.py

```diff
@@ -0,0 +1,497 @@
+
+        if self.name.startswith('fs'):
+            object.__setattr__(self, "dim", 2)
+
+        if self.name in self._standard_spaces:
+            object.__setattr__(self, "standard", True)
+
+        _cohorts = ["%s" % t
+                    for t in _tfapi.TF_LAYOUT.get_cohorts(template=self.name)]
+        if "cohort" in self.spec:
+            if not _cohorts:
+                raise ValueError(
+                    'standard space "%s" does not accept a cohort '
+                    'specification.' % self.name)
+
+            if str(self.spec["cohort"]) not in _cohorts:
+                raise ValueError(
+                    'standard space "%s" does not contain any cohort '
+                    'named "%s".' % (self.name, self.spec["cohort"]))
+        elif _cohorts:
+            _cohorts = ', '.join(['"cohort-%s"' % c for c in _cohorts])
+            raise ValueError(
+                'standard space "%s" is not fully defined.\n'
+                'Set a valid cohort selector from: %s.' % (self.name, _cohorts))
+
+    @property
+    def fullname(self):
+        """Generate a full-name combining cohort."""
+        if "cohort" not in self.spec:
+            return self.name
+        return "%s:cohort-%s" % (self.name, self.spec["cohort"])
+
+    @property
+    def legacyname(self):
+        """Generate a legacy name for fsaverageX spaces."""
+        if self.name == "fsaverage":
+            return FSAVERAGE_LEGACY[self.spec["den"]]
+
+    @name.validator
+    def _check_name(self, attribute, value):
+        if value.startswith('fsaverage'):
+            return
+        valid = list(self._standard_spaces) + NONSTANDARD_REFERENCES
+        if value not in valid:
+            raise ValueError(
+                'space identifier "%s" is invalid.\nValid '
+                'identifiers are: %s' % (value, ', '.join(valid)))
+
+    @spec.validator
+    def _check_spec(self, attribute, value):
+        if value is not None and not isinstance(value, dict):
+            raise TypeError(
+                "invalid space specification: %s." % str(value))
+
+    @classmethod
+    def from_string(cls, value):
+        """Create a new Space from string."""
+        _args = value.split(':')
+        spec = defaultdict(list, {})
+        for modifier in _args[1:]:
+            mitems = modifier.split('-', 1)
+            spec[mitems[0]].append(len(mitems) == 1 or mitems[1])
+
+        allspecs = _expand_entities(spec)
+
+        return [cls(_args[0], s) for s in allspecs]
+
+
+class SpatialReferences:
+    """
+    Manage specifications of spatial references.
+
+    Examples
+    --------
+    >>> sp = SpatialReferences([
+    ...     'func',
+    ...     'fsnative',
+    ...     'MNI152NLin2009cAsym',
+    ...     'anat',
+    ...     'fsaverage5',
+    ...     'fsaverage6',
+    ...     ('MNIPediatricAsym', {'cohort': '2'}),
+    ...     ('MNI152NLin2009cAsym', {'res': 2}),
+    ...     ('MNI152NLin2009cAsym', {'res': 1}),
```

Comment:
Are res `int`s? Can they be `str`?

Reply:
i believe it can be both, similar to density

Reply:
Since the BIDS PR is not merge, these are still `int`s. We will make them `str` when that is official in BIDS-Derivatives.